aggregators/converter.go

```diff
@@ -30,6 +30,213 @@ const (
+	mb.KeyedSpanMetrics[i].Key = &mb.SpanAggregationKey[i]
+	mb.KeyedSpanMetrics[i].Metrics = &mb.SpanMetrics[i]
+	mb.KeyedSpanMetricsSlice = append(mb.KeyedSpanMetricsSlice, &mb.KeyedSpanMetrics[i])
+}
+
+func (p *partitionedMetricsBuilder) addSpanMetrics(e *modelpb.APMEvent, repCount float64) {
+	var key aggregationpb.SpanAggregationKey
+	setSpanKey(e, &key)
+
+	mb := p.get(p.hasher.Chain(&key))
+	i := len(mb.KeyedSpanMetricsSlice)
+	mb.SpanAggregationKey[i] = key
+	setSpanMetrics(e, repCount, &mb.SpanMetrics[i])
+	mb.KeyedSpanMetrics[i].Key = &mb.SpanAggregationKey[i]
+	mb.KeyedSpanMetrics[i].Metrics = &mb.SpanMetrics[i]
+	mb.KeyedSpanMetricsSlice = append(mb.KeyedSpanMetricsSlice, &mb.KeyedSpanMetrics[i])
+}
+
+func (p *partitionedMetricsBuilder) addServiceSummaryMetrics() {
+	// There are no actual metric values, we're just want to
+	// create documents for the dimensions, so we can build a
+	// list of services.
+	_ = p.get(p.hasher)
+}
+
+func (p *partitionedMetricsBuilder) get(h Hasher) *eventMetricsBuilder {
+	partition := p.partitioner.Partition(h)
+	for _, mb := range p.builders {
+		if mb.partition == partition {
+			return mb
+		}
+	}
+	mb := eventMetricsBuilderPool.Get().(*eventMetricsBuilder)
+	mb.partition = partition
+	p.builders = append(p.builders, mb)
+	return mb
+}
+
+// eventMetricsBuilder holds memory for the contents of per-partition
+// ServiceInstanceMetrics. Each instance of the struct is capable
+// of holding as many metrics as may be produced for a single event.
+type eventMetricsBuilder struct {
+	partition uint16
+
+	TransactionHistogramCounts   [1]int64
+	TransactionHistogramBuckets  [1]int32
+	TransactionHistogram         aggregationpb.HDRHistogram
+	TransactionAggregationKey    aggregationpb.TransactionAggregationKey
+	TransactionMetrics           aggregationpb.TransactionMetrics
+	KeyedTransactionMetrics      aggregationpb.KeyedTransactionMetrics
+	KeyedTransactionMetricsArray [1]*aggregationpb.KeyedTransactionMetrics
+	KeyedTransactionMetricsSlice []*aggregationpb.KeyedTransactionMetrics
+
+	ServiceTransactionAggregationKey    aggregationpb.ServiceTransactionAggregationKey
+	ServiceTransactionMetrics           aggregationpb.ServiceTransactionMetrics
+	KeyedServiceTransactionMetrics      aggregationpb.KeyedServiceTransactionMetrics
+	KeyedServiceTransactionMetricsArray [1]*aggregationpb.KeyedServiceTransactionMetrics
+	KeyedServiceTransactionMetricsSlice []*aggregationpb.KeyedServiceTransactionMetrics
+
+	// There can be at most 128 span metrics per event:
+	// - exactly 1 for a span event
+	// - at most 128 (dropped span stats) for a transaction event (1)
+	//
+	// (1) https://github.com/elastic/apm/blob/main/specs/agents/handling-huge-traces/tracing-spans-dropped-stats.md#limits
+	SpanAggregationKey    [128]aggregationpb.SpanAggregationKey
+	SpanMetrics           [128]aggregationpb.SpanMetrics
+	KeyedSpanMetrics      [128]aggregationpb.KeyedSpanMetrics
```

Comment:
This one is subtly different - it's more akin to KeyedTransactionMetrics and KeyedServiceTransactionMetrics. Those are not arrays because there can only be one of them.